### PR TITLE
spec: add "dns" field to the result JSON

### LINKED ISSUE #67: passing dns resolvers out of a plugin
We need to define the where in the returned JSON metadata a plugin can tell the caller what the DNS configuration should be. This will essentially be a priority ordered list of IP addresses:

```
"dns-resolvers": ["8.8.8.8", "8.8.4.4"]
```

### PR DESCRIPTION
**Pull request by eyakubovich**
This adds an ability for networks to communicate
about the DNS nameservers they know about to the
container runtime.

Fixes #67

### CODE REVIEW

**Comment by steveeJ** [comment]
Any particular reason why DNS servers should not be differentiated between IP versions?

**Comment by philips** [comment]
@steveeJ is that common to differentiate by protocol? Can resolv.conf do that?

**Comment by squaremo** [comment]
I can think of three appropriate ways of providing nameservers:

    * entirely out-of-band with respect to CNI (e.g., if Kubernetes runs its own DNS server);
    * in the configuration of a network;
    * supplied by an IPAM plugin.

Allowing that I don't always have the right end of the stick, I would be interested to know what suggests this change, since it doesn't (quite) cover off the above situations.

**Comment by eyakubovich** [comment]
@steveeJ That is a great point and I spent a while thinking about whether to add `dns` at top level or inside the `ip4`, `ip6` sections. The reason I decided to go with top level is to allow the plugin to specify priority/preference between IPv4 and IPv6 nameservers in dual stack scenarios. However, it's a small win so definitely open to doing it the other way if there's a reason.

**Comment by eyakubovich** [comment]
@squaremo I'm not sure I understand your concern. There is no right way to do DNS resolution. This change just allows the CNI plugin to communicate about the nameservers it happened to know about (e.g. b/c it learned it via DHCP). It would still be up to the runtime to choose the implementation.

**Comment by eyakubovich** [comment]
@squaremo I see what you're saying. The IPAM section in the spec is missing the `dns` field. It will ultimately be coming from there and propagated up. I will fix.

**Comment by steveeJ** [comment]
@philips resolv.conf doesn't do it but I wouldn't only solve the resolv.conf use-case in CNI. We provide routes/gateway per protocol, applying that on DNS would be consistent and nice to have for applications that could make use with the extra information.

File: SPEC.md [hunk 1]
```diff
@@ -48,6 +49,7 @@ The operations that the CNI plugin needs to support are:
     - **Name of the interface inside the container**. This is the name that should be assigned to the interface created inside the container (network namespace); consequently it must comply with the standard Linux restrictions on interface names.
   - Result:
     - **IPs assigned to the interface**. This is either an IPv4 address, an IPv6 address, or both.
+    - **List of DNS nameservers**. This is priority ordered list of IPv4 and IPv6 addresses of DNS nameservers.
```
Comment (jonboulle) [inline SPEC.md]: s/priority ordered/a priority-ordered/

File: SPEC.md [hunk 2]
```diff
@@ -83,10 +85,18 @@ Success is indicated by a return code of zero and the following JSON printed to
     "ip": <ipv6-and-subnet-in-CIDR>,
     "gateway": <ipv6-of-the-gateway>,  (optional)
     "routes": <list-of-ipv6-routes>    (optional)
-  }
+  },
+  "dns": <list-of-DNS-nameservers>     (optional)
 }
 ```
 
+The "dns" field contains a list of priority ordered list of DNS nameservers that this network is aware of.
```
Comment (jonboulle) [inline SPEC.md]: s/a list of priority ordered list of/ a priority-ordered list of/

File: SPEC.md [hunk 3]
```diff
@@ -83,10 +85,18 @@ Success is indicated by a return code of zero and the following JSON printed to
...
+Each entry in the list is a string containing either an IPv4 or an IPv6 address.
+Typically this value would just be the value returned by the IPAM plugin.
+It is outside the scope of this specification on how the container runtime uses the list of DNS nameservers from each of the networks to create a final list for the container.
+In addition, it is outside the scope of this specification how the final list is used to provide name resolution services to the container.
```
Comment (jonboulle) [inline SPEC.md]: doesn't this imply line-96 anyway? maybe we could just have one

File: SPEC.md [hunk 4]
```diff
@@ -193,6 +204,10 @@ Each route entry is a dictionary with the following fields:
 - `dst` (string): Destination subnet specified in CIDR notation.
 - `gw` (string): IP of the gateway. If omitted, a default gateway is assumed (as determined by the CNI plugin).
 
+The "dns" field contains a list of priority ordered list of DNS nameservers that this network is aware of.
```
Comment (jonboulle) [inline SPEC.md]: s/a list of priority ordered list of/ a priority-ordered list of/

**Comment by jonboulle** [comment]
Few minor syntactical issues, but other than that this approach lgtm.

File: SPEC.md (hunk 3 above)
Comment (dcbw) [inline SPEC.md]: > doesn't this imply line-96 anyway? maybe we could just have one

Yeah, how about:

"It is outside the scope of this specification how the container runtime uses the list of DNS nameservers from each of the networks to provide name resolution services to the container."

**Comment by dcbw** [comment]
The only scenario I can think of that this wouldn't work for is split DNS, where queries for certain domains are directed to specific nameservers and the rest go somewhere else.  eg, VPN-like functionality.  That would require that the 'dns' option be a map of string::list-of-string though.  And maybe too complicated for now, since if this actually becomes a need we could add a 'split-dns' key or something. Just food for thought.

But generally LGTM too.  There is value in having the list priority ordered, which implies keeping it separate and not in the IP-version-specific blocks.

**Comment by dcbw** [comment]
@rajatchopra look good to you too?

**Comment by rajatchopra** [comment]
Useful and good to go I think.

One question which should not block this PR is handling of /etc/hosts, if at all. Its mention is removed from Open Questions, but this PR does not address it.

**Comment by eyakubovich** [comment]
@rajatchopra This PR takes a stance that `/etc/hosts`, like `/etc/resolve.conf`, is outside the scope of the spec (at least for now). The only thing this spec covers in terms of name resolution is the `dns` key as specified in this PR.

**Comment by eyakubovich** [comment]
@dcbw I think with split-dns, it's the DNS server that serves up alternate IPs for a some DNS zone. It's usually not done by the local resolver (e.g. `/etc/resolv.conf`) so I think it's reasonable not to support it here as well. But like you say, we can extend the spec if the need arises.

**Comment by dcbw** [comment]
> I think with split-dns, it's the DNS server that serves up alternate IPs for a some DNS zone.

@eyakubovich Not really.  It comes into play when you have multiple sources of network information.  For example, suppose you want your container to connect to multiple networks; one for the local cluster and a VPN-like one.  Each network is handled by a different CNI plugin (or the same plugin, doesn't matter really), and each plugin can return whatever DNS information it wants.  Which do you choose to be "primary"?

Even though priority and how DNS resolution is handled is out of scope, the thing that calls the CNI plugin needs some way to figure that out.  And passing the search domains back from the CNI plugin is pretty crucial for split DNS, since that's the only way you know what domains to send to that network's DNS server.

I'm not saying we need to add it now...  Just that we should think about the use-case as more and more stuff becomes multi-network capable.

**Pull request merged**